Add new laser profile: PulseFromSpectrum by @nils-pr

steindev opened the pull request:
Initializes laser pulse with gaussian transverse envelope and longitudinal envelope calculated from complex electric field given in frequency space.

This PR addresses reviewer comments from #3927 which will be closed as the original author @nils-pr is currently unable to provide the requested changes.

From the original PR #3927:
```
New laser profile for PIConGPU.

In this laser profile the laser pulse is defined in the frequency domain, with:

    a Gaussian spectrum
    phase shift for GDD/TOD
    in addition the spectrum and phase is altered to get the spatial shape of a Gaussian beam

Time dependent E-Field is then calculated via Fourier transformation.
Thus a time dependent gaussian beam, with a temporal profile defined by spectrum and freely chosen GDD/TOD,
is described.
```

I added the `CI:no-compile` flag since this laser is not part of any setup we have in the compile suite.

nils-pr committed:
Add new laser PulseFromSpectrum

PrometheusPi commented:
@steindev Is this ready for review or still WIP?

steindev commented:
Ready for review!

steindev commented:
@PrometheusPi would you please review soon? @nils-pr wants to provide a circularly polarized version on top of this.

sbastrakov commented on include/picongpu/fields/laserProfiles/PulseFromSpectrum.hpp:
```diff
@@ -0,0 +1,335 @@
+                        // Norm is choosen so that the maximum after fourier transformation is unity.
+                        float_X norm = math::sqrt(float_X(PI)) * Unitless::PULSE_DURATION;
+                        return norm
+                            * math::exp(
+                                   -1.0
+                                   * math::pow(
+                                       2._X * Unitless::PULSE_DURATION * float_X(PI) * (v - Unitless::v0),
```
Why is this mixed-precision? Perhaps `-math::pow(...)` ?

steindev replied:
Refactored

sbastrakov commented on include/picongpu/fields/laserProfiles/PulseFromSpectrum.hpp:
```diff
@@ -0,0 +1,335 @@
+                                   -1.0
+                                   * math::pow(
+                                       2._X * Unitless::PULSE_DURATION * float_X(PI) * (v - Unitless::v0),
+                                       2.0));
```
If this is critical, could multiply by itself instead of `pow(x, 2)`, same also below

steindev replied:
Refactored

sbastrakov commented on include/picongpu/fields/laserProfiles/PulseFromSpectrum.hpp:
```diff
@@ -0,0 +1,335 @@
+#pragma once
+
+#include "picongpu/simulation_defines.hpp"
+
```
`#include "picongpu/fields/laserProfiles/BaseFunctor.hpp"` is missing

sbastrakov commented on include/picongpu/fields/laserProfiles/PulseFromSpectrum.hpp:
```diff
@@ -0,0 +1,335 @@
+                    {
+                    }
+
+                    /** device side manipulation for init plane (transversal)
```
Is this accurate? This operator appears to do the whole thing

steindev replied:
I don't know. This is copy-paste from the gaussian profile. Any suggestion for improvement? (I actually don't even know what you mean by 'whole thing')

sbastrakov replied:
Yes, it's just outdated in all profiles. The whole thing I meant the whole laser calculation and application

sbastrakov commented on include/picongpu/fields/laserProfiles/PulseFromSpectrum.hpp:
```diff
@@ -0,0 +1,335 @@
+                        offsetToCenterOfPlane[planeNormalDir] = 0; // do not shift origin of plane normal
+                        floatD_X const pos
+                            = precisionCast<float_X>(localCell + offsetToCenterOfPlane) * cellSize.shrink<simDim>();
+                        // @todo add half-cells via traits::FieldPosition< Solver::NumicalCellType, FieldE >()
```
I see this is copied from the existing gaussian profile. We probably can easily account for it in all relevant profiles in a separate PR

steindev replied:
Hmm not sure how useful this would be, for the moment it is ok I think. I also think that we do not want to add a whole bunch of new laser profiles in the future, and if so, then preferably use incidentField.

sbastrakov reviewed (changes requested):
Some small comments above. Also please add the new profile [here](https://github.com/ComputationalRadiationPhysics/picongpu/blob/dev/docs/source/usage/param/laser/profiles.rst) so that it appears in readthedocs

PrometheusPi reviewed (changes requested):
Please adjust also all `laser.param` of all examples that use a laser pulse.

PrometheusPi commented:
Please also setup at least one compile test with this laser in the examples.

PrometheusPi commented:
@steindev What is the status oof this pull request?

steindev commented:
> Please also setup at least one compile test with this laser in the examples.

As we just found out, we don't have compile time test for lasers yet :sweat_smile:

steindev commented:
Addressed reviewer comments! Ready for re-review!

PrometheusPi commented:
Thanks @steindev 
I will have a look as soon as possible.

steindev commented:
@PrometheusPi @sbastrakov knock, knock.

PrometheusPi commented:
@steindev I will have a look at this pull request today.

steindev commented:
**EDIT:** Forget about this message. See next message.

@PrometheusPi I encountered a couple of errors in the implementation and fixed these now (latest force-push). A comparison between a normal gaussian beam initialized via PulseFromSpectrum and one initialized via our standard GaussianBeam yields identical results now. Result is attached.
[laser-from-spectrum_validaton.xcf.zip](https://github.com/ComputationalRadiationPhysics/picongpu/files/8421870/laser-from-spectrum_validaton.xcf.zip)

Please re-review.

steindev commented:
Ohh :cursing_face:, in my test I ran the standard gauss laser again instead of PulseFromSpectrum. The latter still does not work. Will debug today :disappointed:

steindev commented:
Finally! The errors I introduced and errors in the original implementation are now fixed. See attached gimp file for comparison between our standard gauss laser in PIConGPU and PulseFromSpectrum without dispersion. Toggle visibility of the layers: What is in the top layer needs to be applied to the bottom layer(=gauss reference) in order to obtain the middle layer(=PulseFromSpectrum) => Excellent agreement!
[laser-from-spectrum_implementation-differences.xcf.zip](https://github.com/ComputationalRadiationPhysics/picongpu/files/8476680/laser-from-spectrum_implementation-differences.xcf.zip)

Thanks to @sbastrakov for the help!

@PrometheusPi please re-review and merge if there are no more comments.

sbastrakov commented:
Is the approach implemented in this PR somehow similar to what SMILEI does https://smileipic.github.io/Smilei/laser_offset.html#laser-propagation-preprocessing ?

steindev commented:
> Is the approach implemented in this PR somehow similar to what SMILEI does https://smileipic.github.io/Smilei/laser_offset.html#laser-propagation-preprocessing ?

Well, there is definitely a similarity in the sense that we have the spectrum over ω and calculate the field in time domain from this spectrum. However, there are two notable differences:
(1) We don't do the propagation from an arbitrary point to the initialization plane. Yet the "propagation" they do seems rather useless to me as they just shift the phase a little instead of performing a real propagation using the Kirchhoff-Fresnel diffraction integral. Their approach is only valid in the very near surrounding of the position where the field is known and I actually doubt that it results in a real difference on the final field in the simulation. But I might overlook something here. Employing the Kirchhoff-Fresnel diffraction integral would allow propagation by (almost) arbitrarily long distances. "Almost" refers to the fact that the propagation distance must be at least a few wavelengths.
(2) We already assume a spectrum which can be evaluated via a function call (either as a given analytic function such as a Gaussian or an interpolating polynomial of a more complex spectral shape), while they generate the spectrum by sampling and Fourier transforming a given temporal field evolution. In essence this is equal, as their spectrum obtained through Fourier transformation can be fitted by a function and this function is what we would use as spectrum. In principle, they save the user from performing this Fourier transform and fit by himself and their invest is additional memory to save the spectrum. Then again, we can measure the frequency spectrum directly in the lab (which is done to reconstruct the temporal laser pulse shape in a single shot) and fit it and put it in the simulation.

steindev commented:
ping @PrometheusPi. Merge! Merge! Merge! (It is tested and nothing to be afraid of!) :sunglasses:

sbastrakov commented:
@steindev thanks for explanation.  With my newly obtained knowledge of incident field at angle I may now also understand the SMILEI approach better than at time of sharing that link. Easier to discuss offline

psychocoderHPC commented:
@PrometheusPi please check if your points are addressed!

PrometheusPi commented on include/picongpu/param/laser.param:
```diff
@@ -604,6 +606,120 @@ namespace picongpu
+                enum PolarisationType
+                {
+                    LINEAR_X = 1u,
+                    LINEAR_Z = 2u,
```
why is there no circular polarized option?

steindev replied:
@nils-pr wanted to add it once this PR is merged. But this is quite some time ago. Not sure if he still wants to do it.

PrometheusPi commented on include/picongpu/fields/laserProfiles/PulseFromSpectrum.hpp:
```diff
@@ -0,0 +1,330 @@
+                    float_64 const waist = Unitless::W0 * math::sqrt(1.0 + y_rel * y_rel);
+
+                    float_64 envelope = float_64(Unitless::AMPLITUDE);
+                    if(simDim == DIM2)
+                        envelope *= math::sqrt(float_64(Unitless::W0) / waist);
+                    else if(simDim == DIM3)
+                        envelope *= float_64(Unitless::W0) / waist;
```
Do we do this for all lasers? 
What is the effect of this multiplication?

steindev replied:
Yes, it is required as the laser amplitude scales different with distance from the focus for two or three dimensional propagation.

PrometheusPi reviewed (changes requested):
I ran the default LWFA setup with the following changes:
 - no particles
 - X polarized laser
 and compared it to the new spectral laser using the same parameters. Despite using the same a_0, w_0, tau, and pulse init - the laser pulse had a different position after sliding. Example plots for time step 4000 are given below. 
 
 Field energy looks fine - Gauss and (Gauss-like) spectral laser agree (after initialization).
 
 My guess is, that there is something different in how `PULSE_INIT` is defined or works. 
 @steindev If you say, there should be no perfect agreement between `GaussianBeam` and `PulseFromSpectrum` if GDD and TOD  are zero - I am fine with merging.

PrometheusPi commented:
`GaussianBeam`
<img width="1564" alt="spectral" src="https://user-images.githubusercontent.com/5121158/167320352-edee2ce0-1293-4410-ba7a-5def1648f64c.png">


`PulseFromSpectrum`
![grafik](https://user-images.githubusercontent.com/5121158/167320459-df200ec1-6669-4b11-8cd3-e14f15ebf981.png)

PrometheusPi commented:
I see that my resizing of the image caused the shift in init position - thus everything is fine!

PrometheusPi reviewed (approved):
No shift in init duration, energy etc. agrees.
